### DISCUSSION TITLE
EVAL : Grid search (bruteforce) F1 + IoU threshold

### DISCUSSION THREAD

**Discussion by ExtReMLapin** [ultralytics/ultralytics#21494] [General]
At evaluation end, it performs eval to find best conf F1 score.

But in our case, it may be useful to also adjust IoU threshold that outputs a heatmap and best possible score.

From my understanding, by default IoU is set a 0.5 which sounds fair in most cases.
<img width="1920" height="1440" alt="f1_heatmap" src="https://github.com/user-attachments/assets/9a423895-db85-4d54-be3a-4bf5afa450c4" />

```python
import numpy as np
import matplotlib.pyplot as plt
from ultralytics import YOLO
from ultralytics.utils.metrics import ConfusionMatrix
from ultralytics.data.utils import check_det_dataset
from pathlib import Path
import cv2
import torch

model = YOLO("YOURMODEL.pt")


data_dict = check_det_dataset("YOUR_DATASET.yaml")
val_dir = Path(data_dict["val"])


images = list(val_dir.rglob("*.jpg"))  

#images = images[:100]



preds_all = []
labels_all = []
import tqdm
for img_path in tqdm.tqdm(images):
    img = cv2.imread(str(img_path))
    results = model(img, conf=0.001, max_det=1500, verbose=False)[0]
    preds_all.append(results.boxes.data.cpu())
    
    label_path = (val_dir.parent.parent / "labels" / "val" / (img_path.stem + ".txt"))
    labels_tensor = []
    if label_path.exists():
        with open(label_path) as f:
            for line in f:
                cls, x, y, w, h = map(float, line.split())
                
                H, W = img.shape[:2]
                x1 = (x - w/2) * W
                y1 = (y - h/2) * H
                x2 = (x + w/2) * W
                y2 = (y + h/2) * H
                labels_tensor.append([0, x1, y1, x2, y2])  
    else:
        print("label path doesn't exist", label_path)
    labels_all.append(np.array(labels_tensor))




conf_thresholds = np.linspace(0, 1, 100)
iou_thresholds = np.linspace(0.001, 0.99, 100)
f1_matrix = np.zeros((len(iou_thresholds), len(conf_thresholds)))
recall_matrix = np.zeros((len(iou_thresholds), len(conf_thresholds)))
precision_matrix =  np.zeros((len(iou_thresholds), len(conf_thresholds)))


from concurrent.futures import ProcessPoolExecutor, as_completed
import os, torch, numpy as np, tqdm

torch.set_num_threads(1)  
NC = len(model.names)

def _init_worker(_preds, _labels):
    global preds_all_g, labels_all_g
    preds_all_g, labels_all_g = _preds, _labels

def _eval_cell(i, j, iou_thr, conf_thr):
    cm = ConfusionMatrix(nc=NC, conf=conf_thr, iou_thres=iou_thr)
    for pred, label in zip(preds_all_g, labels_all_g):
        if len(label):
            gt_bboxes = torch.tensor(label[:, 1:5], dtype=torch.float32)
            gt_cls = torch.tensor(label[:, 0], dtype=torch.int64)
        else:
            gt_bboxes = torch.zeros((0, 4))
            gt_cls = torch.zeros((0,), dtype=torch.int64)
        cm.process_batch(pred, gt_bboxes, gt_cls)

    tp, fp = cm.tp_fp()
    fn = cm.matrix.sum(0)[:-1] - tp
    precision = tp / (tp + fp + 1e-9)
    recall = tp / (tp + fn + 1e-9)
    f1 = 2 * precision * recall / (precision + recall + 1e-9)
    return i, j, np.nanmean(f1), recall, precision

tasks = [(i, j, iou_thr, conf_thr)
         for i, iou_thr in enumerate(iou_thresholds)
         for j, conf_thr in enumerate(conf_thresholds)]

with ProcessPoolExecutor(max_workers=os.cpu_count(),
                         initializer=_init_worker,
                         initargs=(preds_all, labels_all)) as ex:
    futures = [ex.submit(_eval_cell, *t) for t in tasks]
    for fut in tqdm.tqdm(as_completed(futures), total=len(futures), desc="grid"):
        i, j, f1, rec, prec = fut.result()
        f1_matrix[i, j] = f1
        recall_matrix[i, j] = rec
        precision_matrix[i, j] = prec



plt.imshow(f1_matrix, aspect='auto', extent=[0,1,iou_thresholds[0],iou_thresholds[-1]], origin='lower')
plt.colorbar(label="F1")
plt.xlabel("Confidence")
plt.ylabel("IoU Threshold")
plt.title("F1 Heatmap")
plt.tight_layout()
plt.savefig("f1_heatmap.png", dpi=300)
plt.close()




best_idx = np.unravel_index(np.argmax(f1_matrix), f1_matrix.shape)
best_iou = iou_thresholds[best_idx[0]]
best_conf = conf_thresholds[best_idx[1]]
best_f1 = f1_matrix[best_idx]

print(f"Best IoU: {best_iou:.2f}, Best Confidence: {best_conf:.2f}, F1: {best_f1:.4f}")


import json


heatmap_data = {
    "f1_matrix": f1_matrix.tolist(),
    "recall_matrix": recall_matrix.tolist(),
    "precision_matrix": precision_matrix.tolist(),
    "conf_thresholds": conf_thresholds.tolist(),
    "iou_thresholds": iou_thresholds.tolist(),
    "best_results": {
        "iou": float(best_iou),
        "confidence": float(best_conf),
        "f1_score": float(best_f1)
    },
    "f1_matrix_shape": f1_matrix.shape,
    "recall_matrix_shape" : recall_matrix.shape,
    "precision_matrix_shape" : precision_matrix.shape
}


with open("f1_heatmap_data.json", "w") as f:
    json.dump(heatmap_data, f, indent=2)

print("Heatmap data exported to f1_heatmap_data.json")
```

**Comment by UltralyticsAssistant** [comment]
👋 Hello @ExtReMLapin, thank you for sharing your detailed workflow and suggestions with the Ultralytics community! 🚀 Your exploration of grid searching over both confidence and IoU thresholds and visualizing the F1 heatmap is impressive and could be very helpful for users fine-tuning evaluation metrics.

This is an automated response to help guide your discussion while an Ultralytics engineer reviews your post and provides further assistance soon.

For reference:
- If this is a 🐛 Bug Report, please include a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us investigate the issue efficiently.
- If you have a custom training ❓ Question, please provide more details such as dataset samples and training logs, and review our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Explore our [Docs](https://docs.ultralytics.com/) for more [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples, which may address some aspects of your inquiry.

Join the Ultralytics community where it suits you best:
- For real-time chat, visit [Discord](https://discord.com/invite/ultralytics) 🎧
- For in-depth discussions, check out [Discourse](https://community.ultralytics.com/)
- Or participate in threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/)

## Upgrade

To ensure you have the latest features and fixes, upgrade to the most recent `ultralytics` package (and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml)) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in various up-to-date, verified environments (with all dependencies like [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests cover all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thanks again for your input! An Ultralytics engineer will review your discussion and follow up soon.

**Reply by ExtReMLapin** [reply]
How can I make this bot SHUT THE FUCK HELL UP
